File: algorithm/341.扁平化嵌套列表迭代器.ts

// This is the interface that allows for creating nested lists.
// You should not implement it, or speculate about its implementation
class NestedInteger {
  // If value is provided, then it holds a single integer
  // Otherwise it holds an empty nested list
  constructor(value?: number) {
    // ...
  };

  // Return true if this NestedInteger holds a single integer, rather than a nested list.
  // @ts-ignore
  isInteger(): boolean {
    // ...
  };

  // Return the single integer that this NestedInteger holds, if it holds a single integer
  // Return null if this NestedInteger holds a nested list
  // @ts-ignore
  getInteger(): number | null {
    // ...
  };

  // Set this NestedInteger to hold a single integer equal to value.
  setInteger(value: number) {
    // ...
  };

  // Set this NestedInteger to hold a nested list and adds a nested integer elem to it.
  add(elem: NestedInteger) {
    // ...
  };

  // Return the nested list that this NestedInteger holds,
  // or an empty list if this NestedInteger holds a single integer
  // @ts-ignore
  getList(): NestedInteger[] {
    // ...
  };
};


class NestedIterator {
  private flattenList: NestedInteger[];
  constructor(nestedList: NestedInteger[]) {
    this.flattenList = this.initFlattenList(nestedList);
  }

  initFlattenList = (list: NestedInteger[]): NestedInteger[] => (list.reduce((a, b) => (b.isInteger() ? b : this.initFlattenList(b.getList()) as any), []));

  hasNext(): boolean {
    return !!this.flattenList.length;
  }

  next(): number | null {
    return this.flattenList.shift().getInteger();
  }
}

/**
* Your ParkingSystem object will be instantiated and called as such:
* var obj = new NestedIterator(nestedList)
* var a: number[] = []
* while (obj.hasNext()) a.push(obj.next());
*/

File: algorithm/341.扁平化嵌套列表迭代器.cpp
#include "iostream"
#include "string"
#include "vector"
#include "algorithm"
#include "stack"
#include "queue"
using namespace std;

class NestedInteger {
public:
  // Return true if this NestedInteger holds a single integer, rather than a nested list.
  bool isInteger() const;

  // Return the single integer that this NestedInteger holds, if it holds a single integer
  // The result is undefined if this NestedInteger holds a nested list
  int getInteger() const;

  // Return the nested list that this NestedInteger holds, if it holds a nested list
  // The result is undefined if this NestedInteger holds a single integer
  const vector<NestedInteger> &getList() const;
};

class Command {
  public:
    string cmd = "";
    vector<NestedInteger> ni;
    NestedInteger interger;
    Command(string cmd, vector<NestedInteger> ni): cmd(cmd), ni(ni) {};
    Command(string cmd, NestedInteger interger): cmd(cmd), interger(interger) {};
};

class NestedIterator {
private:
  queue<NestedInteger> flattenList;
public:
  NestedIterator(vector<NestedInteger> &nestedList) {
    if (nestedList.size()) {
      stack<Command*> myStack;
      NestedInteger one = nestedList[0];
      Command *head = new Command("goto", nestedList);
      myStack.push(head);
      while(myStack.size()) {
        Command *command = myStack.top();
        myStack.pop();
        if (command -> cmd == "push") {
          this -> flattenList.push(command -> interger);
        } else {
          if (command -> cmd == "goto") {
            vector<NestedInteger> temp = command -> ni;
            for (int i = temp.size() - 1; i >= 0; i--) {
              Command *cmd;
              if (temp[i].isInteger()) {
                cmd = new Command("push", temp[i]);
              } else {
                cmd = new Command("goto", temp[i].getList());
              }
              myStack.push(cmd);
            }
          }
        }
      }
    }
  }
  
  int next() {
    NestedInteger temp = this -> flattenList.front();
    this -> flattenList.pop();
    return temp.getInteger();
  }
  
  bool hasNext() {
    return !!flattenList.size();
  }
};

/**
 * Your NestedIterator object will be instantiated and called as such:
 * NestedIterator i(nestedList);
 * while (i.hasNext()) cout << i.next();
 */